### PR TITLE
Replace matplotlib plot config dialog

### PR DESCRIPTION
**Pull request by ghost**
**This depends on:** #25909, which should be merged first. 

**Description of work.**
The matplotlib options dialog has been replaced by a similar dialog offering more options. The dialog is opened in a similar way, using the cog button on a figure.

There are three tabs within the main dialog: Axes, Curves (meaning Line2D or ErrorbarContainer objects) and Images. These tabs will only be created if there are axes, curves or images respectively in the figure. To update curves we remove the old and re-plot them. For images we can set all the properties we need to without re-plotting.

Axes tab:
- Change axes' titles, scales and labels.

Curves tab:
- Hide a curve, set a curve's label or delete a curve
- Contains 3 sub-tabs: Line, Marker and Errorbars.
- The Errorbars tab will only be enabled if the curve's workspace has errorbars, or a scripted plot has errorbars. There is an option to hide/show these bars.

Images:
- For `imshow` or `colormesh` plots
- You can now set a logarithmic scaling on a colorfill
- Ensure if you're only plotting lines or errorbars the Curves tab is visible and the Images tab is not

**Note**: there are plans to add some more features to this issue, but this is quite a big PR already and it now offers at least all the options that were provided with the default matplotlib dialog, so can be merged. There are plans to add:
- A legend tab to customise the look of the legend (font size, transparency etc.)
- An "Apply to All" button to apply curve/image properties to all curves/images

**To test:**
- Run tests within the `mantid\qt\python\mantidqt\widgets\plotconfigdialog\` directory.
- Plot some figures from workspaces and also via the script window. Make sure the options dialog behaves as you would expect.
- Make sure other plot features work as expected after using the options dialog. 
    - right-clicking the plot and toggling normalize by bin width
    - right-cliking and toggling axis scales
    - deleting a workspace removes the associated curves from a plot

Fixes (partially) #25550 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/python/mantidqt/widgets/plotconfigdialog/__init__.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+#  This file is part of the mantid workbench.
+
+from matplotlib.axes import ErrorbarContainer
+from matplotlib.collections import QuadMesh
+
+
+def generate_ax_name(ax):
+    """
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/plotconfigdialog/__init__.py]: Great method docs :+1:

File: qt/python/mantidqt/widgets/plotconfigdialog/__init__.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI,
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+#  This file is part of the mantid workbench.
+
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/plotconfigdialog/__init__.py]: Could you check all of the files and make sure they all at least have the `from __future__ import (absolute_import, division, unicode_literals)` imports?

**Review by martyngigg — CHANGES_REQUESTED** [review]
I got a problem starting the dialog under Python 3:
```
Traceback (most recent call last):
  File "/media/data1/source/github/mantidproject/mantid/qt/applications/workbench/workbench/plotting/figuremanager.py", line 291, in launch_plot_options
    parent=self.window)
  File "/media/data1/source/github/mantidproject/mantid/qt/python/mantidqt/widgets/plotconfigdialog/presenter.py", line 32, in __init__
    axes_tab = AxesTabWidgetPresenter(self.fig, parent=self.view)
  File "/media/data1/source/github/mantidproject/mantid/qt/python/mantidqt/widgets/plotconfigdialog/axestabwidget/presenter.py", line 30, in __init__
    self.update_view()
  File "/media/data1/source/github/mantidproject/mantid/qt/python/mantidqt/widgets/plotconfigdialog/axestabwidget/presenter.py", line 84, in update_view
    self.view.set_title(ax_props.title)
  File "/media/data1/source/github/mantidproject/mantid/qt/python/mantidqt/widgets/plotconfigdialog/axestabwidget/view.py", line 54, in set_title
    self.axes_title_line_edit.setText(title)
TypeError: QLineEdit.setText(str): argument 1 has unexpected type 'bytes'
```
There is a function in `mantid.py3compat` called `ensure_str` that might help?

Testing under Python 2 on Ubuntu 16.04 I can see the dialog but the first thing I notice is that the sizing is not quite right. The text is cut off in most boxes and I can't seem to resize the dialog at all.

The options generally seem to work well as far as I can see them though with a quick play. I thought it best to get these first batch of comments up quickly.

**Comment by martyngigg** [comment]
I have unassigned myself as I'm now away for a week.

File: Framework/PythonInterface/mantid/plots/__init__.py [hunk 3]
```diff
@@ -300,6 +311,23 @@ def check_axes_distribution_consistency(self):
                 logger.warning("You are overlaying distribution and "
                                "non-distribution data!")
 
+    def artist_has_errorbars(self, artist):
+        """Check if the given artist's workspace has errors"""
+        workspace, spec_num = self.get_artists_workspace_and_spec_num(artist)
+        workspace_index = workspace.getIndexFromSpectrumNumber(spec_num)
+        if any(workspace.readE(workspace_index) != 0):
```
[the remark this answers is not in the document]
Comment (hsautessella) [reply Framework/PythonInterface/mantid/plots/__init__.py]: This is a badly named function. It's checking if the *workspace* associated with the artist has errors. This is used to decide whether the "Errorbars" tab is enabled in the settings for a given artist. I'll rename it

**Review by dtasev — CHANGES_REQUESTED** [review]
Test workspace:
```
ws2d_histo = CreateWorkspace(DataX=[10, 20, 30, 10, 20, 30, 10, 20, 30],
                             DataY=[2, 3, 4, 5, 3, 5],
                             DataE=[1, 2, 3, 4, 1, 1],
                             NSpec=3,
                             Distribution=True,
                             UnitX='Wavelength',
                             VerticalAxisUnit='DeltaE',
                             VerticalAxisValues=[4, 6, 8])
```
The following are done on `ws2d_histo`!

------

#### `Hide Errorbars` seem to 'prefer' being unchecked. To replicate:
- Plot spectrum
- Click options wheel > Curves > Errorbars > Uncheck `Hide` > Click OK
- Go back and hide them again
- Open the options window and go to errorbars, again, this time they will still be unchecked
  - This causes the errorbars to always enable themselves on their own, if the options window is just opened and closed without changing anything

#### Opening figure options and clicking Apply (without doing anything) moves the label to the bottom left - is this intended?

#### Updating labels on an axis doesn't seem to change the text on the axis
- This works if I double click on the axis label, outside of the window

#### Errorbars for spectra 1 turns on on its own
- Open Figure options, click OK without doing anything
- Open again, click OK again, without doing anything
- Spectra 1 should have errors on

------------
```
ws2d_1spec = CreateWorkspace(DataX=[10, 20, 30],
                             DataY=[2, 3],
                             DataE=[1, 2],
                             NSpec=1,
                             Distribution=True,
                             UnitX='Wavelength',
                             VerticalAxisUnit='DeltaE',
                             VerticalAxisValues=[4])
```

The following is on `ws2d_1spec`!

#### Toggle error bars on - the limits are outside of the visible plot.
- An approach like [this one](https://github.com/mantidproject/mantid/pull/25617/files#diff-59a3f22cce9ee9b0ef09d3d8db9c073dR384) might be necessary (change the limit, with a margin). Original comment [here](https://github.com/mantidproject/mantid/pull/25617#discussion_r293433743)

#### Removing last curve and clicking `Ok` shows uncaught exception
- Not sure how to handle this one - forbid removal of last curve?

**Comment by hsautessella** [comment]
I've added some changes that should fix all the issues raised above, as well as some more tests to check that properties are being properly set. 

I've overriden the `relim` method on `MantidAxes` so that it takes errorbars into account when setting axes limits. 

I've also started storing the latest set of properties for curves and axes in the presenters; if that latest set is the same as the the set of properties in the view, we don't apply those properties (I'm in the process of writing up tests for this, but I thought it would be best to get these changes looked at asap). Storing these properties was required for autoscaling, since we were setting the axes limits from the values in the view which was overriding the autoscale. Not re-setting properties that haven't been changed is probably a good idea anyway.

**Review by dtasev — APPROVED** [review]
All the issues from my preview review have been fixed. It works well, code looks very good with nice test coverage.

I think this is ready to be merged in, and for https://github.com/mantidproject/mantid/pull/25617 to be re-based to use the work from this, which should make it a lot faster finish

**Review by martyngigg — APPROVED** [review]

**Pull request merged**